mlx/transforms.cpp

```diff
@@ -216,13 +218,43 @@ void eval(const std::vector<array>& outputs) {
   auto synchronizer =
       array({}, bool_, std::make_unique<Synchronizer>(stream), outputs);
 
-  recurse = [&](const array& a) {
+  compute_depths = [&](const array& a) {
+    auto id = a.id();
+    if (auto it = depth_map.find(id); it != depth_map.end()) {
+      return it->second;
+    }
+    int depth = 0;
+    for (auto& in : a.inputs()) {
+      depth = std::max(depth, compute_depths(in));
+    }
+    depth++;
+    depth_map[id] = depth;
+    for (auto& s : a.siblings()) {
+      depth_map[s.id()] = depth;
+    }
+    return depth;
+  };
+
+  recurse = [&](const array& a, bool largest_branch_first) {
```

Comment:
Most of our primitives don't have too many inputs as of yet, so the process to build the tape in eval should be fine for that - but it might be nice to profile the time spent there in some common use cases just to be sure